# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/fetch/FetchService.h
# Commit: 66b279e5a513
# Full Hash: 66b279e5a513bf706129ac64cf44e3353e3a0d71
# Author: Eden Chuang <echuang@mozilla.com>
# Date: 2023-01-12 21:30:33
# Regressor Bug: 1351231
# File Overlap Count: 1
# Description:
#   Bug 1351231 - FetchService integration for PFetch. r=dom-worker-reviewers,jesup
#   
#   Depends on D142436
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D142437
# ==============================================================================

diff -r 3eb8fdd0ba6d -r 66b279e5a513 dom/fetch/FetchService.h
--- a/dom/fetch/FetchService.h	Thu Jan 12 17:24:28 2023 +0000
+++ b/dom/fetch/FetchService.h	Thu Jan 12 17:24:28 2023 +0000
@@ -14,6 +14,8 @@
 #include "mozilla/dom/FetchTypes.h"
 #include "mozilla/dom/PerformanceTimingTypes.h"
 #include "mozilla/dom/SafeRefPtr.h"
+#include "mozilla/ipc/PBackgroundSharedTypes.h"
+#include "mozilla/net/NeckoChannelParams.h"
 
 class nsILoadGroup;
 class nsIPrincipal;
@@ -24,6 +26,8 @@
 
 class InternalRequest;
 class InternalResponse;
+class ClientInfo;
+class ServiceWorkerDescriptor;
 
 using FetchServiceResponse = SafeRefPtr<InternalResponse>;
 using FetchServiceResponseAvailablePromise =
@@ -72,6 +76,29 @@
   NS_DECL_ISUPPORTS;
   NS_DECL_NSIOBSERVER;
 
+  struct NavigationPreloadArgs {
+    SafeRefPtr<InternalRequest> mRequest;
+    nsCOMPtr<nsIChannel> mChannel;
+  };
+
+  struct WorkerFetchArgs {
+    SafeRefPtr<InternalRequest> mRequest;
+    mozilla::ipc::PrincipalInfo mPrincipalInfo;
+    nsCString mWorkerScript;
+    Maybe<ClientInfo> mClientInfo;
+    Maybe<ServiceWorkerDescriptor> mController;
+    Maybe<net::CookieJarSettingsArgs> mCookieJarSettings;
+    bool mNeedOnDataAvailable;
+    nsCOMPtr<nsICSPEventListener> mCSPEventListener;
+    nsCOMPtr<nsISerialEventTarget> mEventTarget;
+    nsID mActorID;
+  };
+
+  struct UnknownArgs {};
+
+  using FetchArgs =
+      Variant<NavigationPreloadArgs, WorkerFetchArgs, UnknownArgs>;
+
   static already_AddRefed<FetchService> GetInstance();
 
   static RefPtr<FetchServicePromises> NetworkErrorResponse(nsresult aRv);
@@ -80,10 +107,9 @@
 
   // This method creates a FetchInstance to trigger fetch.
   // The created FetchInstance is saved in mFetchInstanceTable
-  RefPtr<FetchServicePromises> Fetch(SafeRefPtr<InternalRequest> aRequest,
-                                     nsIChannel* aChannel = nullptr);
+  RefPtr<FetchServicePromises> Fetch(FetchArgs&& aArgs);
 
-  void CancelFetch(RefPtr<FetchServicePromises>&& aPromises);
+  void CancelFetch(const RefPtr<FetchServicePromises>&& aPromises);
 
  private:
   /**
@@ -94,22 +120,15 @@
    * FetchInstance triggers fetch by instancing a FetchDriver with proper
    * initialization. The general usage flow of FetchInstance is as follows
    *
-   * RefPtr<FetchInstance> fetch = MakeRefPtr<FetchInstance>(aResquest);
-   * fetch->Initialize();
+   * RefPtr<FetchInstance> fetch = MakeRefPtr<FetchInstance>();
+   * fetch->Initialize(FetchArgs args);
    * RefPtr<FetchServicePromises> fetch->Fetch();
    */
   class FetchInstance final : public FetchDriverObserver {
    public:
-    explicit FetchInstance(SafeRefPtr<InternalRequest> aRequest);
+    FetchInstance() = default;
 
-    // This method is used for initialize the fetch.
-    // It accepts a nsIChannel for initialization, this is for the navigation
-    // preload case since there has already been an intercepted channel for
-    // dispatching fetch event, and needed information can be gotten from the
-    // intercepted channel.
-    // For other case, the fetch needed information be created according to the
-    // mRequest
-    nsresult Initialize(nsIChannel* aChannel = nullptr);
+    nsresult Initialize(FetchArgs&& aArgs);
 
     RefPtr<FetchServicePromises> Fetch();
 
@@ -125,17 +144,18 @@
     void FlushConsoleReport() override;
 
    private:
-    ~FetchInstance();
+    ~FetchInstance() = default;
 
     SafeRefPtr<InternalRequest> mRequest;
     nsCOMPtr<nsIPrincipal> mPrincipal;
     nsCOMPtr<nsILoadGroup> mLoadGroup;
     nsCOMPtr<nsICookieJarSettings> mCookieJarSettings;
     RefPtr<PerformanceStorage> mPerformanceStorage;
+    FetchArgs mArgs{AsVariant(FetchService::UnknownArgs())};
     RefPtr<FetchDriver> mFetchDriver;
     SafeRefPtr<InternalResponse> mResponse;
-
     RefPtr<FetchServicePromises> mPromises;
+    bool mIsWorkerFetch{false};
   };
 
   ~FetchService();